message/validation/consensus_validation.go

```diff
@@ -0,0 +1,361 @@
+				return ErrDuplicatedProposalWithDifferentData
+			}
+
+			limits := maxMessageCounts(len(share.Committee), int(share.Quorum))
+			if err := signerState.MessageCounts.Validate(msg, limits); err != nil {
+				return err
+			}
+		}
+	}
+
+	if err := mv.validateJustifications(share, signedMsg); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (mv *MessageValidator) validateDutiesCount(state *SignerState, role spectypes.BeaconRole) error {
+	switch role {
+	case spectypes.BNRoleAttester, spectypes.BNRoleAggregator, spectypes.BNRoleValidatorRegistration:
+		if state.EpochDuties > maxDutiesPerEpoch {
+			err := ErrTooManyDutiesPerEpoch
+			err.got = fmt.Sprintf("%v (role %v)", state.EpochDuties, role)
+			err.want = maxDutiesPerEpoch
+			return err
+		}
+		// TODO: check other roles
+	}
+
+	return nil
+}
+
+func (mv *MessageValidator) hasFullData(signedMsg *specqbft.SignedMessage) bool {
+	return (signedMsg.Message.MsgType == specqbft.ProposalMsgType ||
+		signedMsg.Message.MsgType == specqbft.RoundChangeMsgType ||
+		mv.isDecidedMessage(signedMsg)) && len(signedMsg.FullData) != 0 // TODO: more complex check of FullData
+}
+
+func (mv *MessageValidator) isDecidedMessage(signedMsg *specqbft.SignedMessage) bool {
+	return signedMsg.Message.MsgType == specqbft.CommitMsgType && len(signedMsg.Signers) > 1
+}
+
+func (mv *MessageValidator) maxRound(role spectypes.BeaconRole) specqbft.Round {
+	switch role {
+	case spectypes.BNRoleAttester, spectypes.BNRoleAggregator: // TODO: check if value for aggregator is correct as there are messages on stage exceeding the limit
+		return 12 // TODO: consider calculating based on quick timeout and slow timeout
+	case spectypes.BNRoleProposer, spectypes.BNRoleSyncCommittee, spectypes.BNRoleSyncCommitteeContribution:
+		return 6
+	case spectypes.BNRoleValidatorRegistration:
+		return 0
+	default:
+		panic("unknown role")
+	}
+}
+
+func (mv *MessageValidator) currentEstimatedRound(sinceSlotStart time.Duration) specqbft.Round {
+	if currentQuickRound := specqbft.FirstRound + specqbft.Round(sinceSlotStart/roundtimer.QuickTimeout); currentQuickRound <= roundtimer.QuickTimeoutThreshold {
+		return currentQuickRound
+	}
+
+	sinceFirstSlowRound := sinceSlotStart - (time.Duration(roundtimer.QuickTimeoutThreshold) * roundtimer.QuickTimeout)
+	estimatedRound := roundtimer.QuickTimeoutThreshold + specqbft.FirstRound + specqbft.Round(sinceFirstSlowRound/roundtimer.SlowTimeout)
+	return estimatedRound
+}
+
+func (mv *MessageValidator) waitAfterSlotStart(role spectypes.BeaconRole) time.Duration {
+	switch role {
+	case spectypes.BNRoleSyncCommitteeContribution:
+		return mv.netCfg.Beacon.SlotDurationSec() // TODO: do we need this?
+	case spectypes.BNRoleAttester, spectypes.BNRoleAggregator, spectypes.BNRoleSyncCommittee, spectypes.BNRoleProposer, spectypes.BNRoleValidatorRegistration:
+		return 0
+	default:
+		panic("unknown role")
+	}
+}
```

Comment:
if this won't be used outside of tests, i think it's better to move this to the test file (as a func that requires `netCfg` for example)

Reply:
It was unused because I disabled the estimated round check, but now it's used again